pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,774 @@
+	klog.V(2).InfoS("Uploading captured packets for PacketCapture", "name", pc.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload packets while getting uploader: %w", err)
+	}
+	authSecret := v1.SecretReference{
+		Name:      fileServerAuthSecretName,
+		Namespace: env.GetAntreaNamespace(),
+	}
+	serverAuth, err := ftp.ParseFileServerAuth(ftp.BasicAuthentication, &authSecret, c.kubeClient)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get authentication for the fileServer", "name", pc.Name, "authSecret", authSecret)
+		return err
+	}
+	cfg := ftp.GenSSHClientConfig(serverAuth.BasicAuthentication.Username, serverAuth.BasicAuthentication.Password)
+	return uploader.Upload(pc.Spec.FileServer.URL, c.generatePacketsPathForServer(pc.Name), cfg, outputFile)
+}
+
+func isCaptureCompleted(pc *crdv1alpha1.PacketCapture, num int32) bool {
+	cfg := pc.Spec.CaptureConfig.FirstN
+	if cfg != nil && num == cfg.Number {
+		return true
+	}
+	return false
+}
+
+func (c *Controller) patchStatusIfNeeded(name string, status packetCapturePhase) error {
+	klog.InfoS("Patching PacketCapture CR Status if needed", "name", name, "status", status)
+	conditions := []crdv1alpha1.PacketCaptureCondition{}
+
+	if status == packetCapturePhaseRunning {
+		conditions = append(conditions, crdv1alpha1.PacketCaptureCondition{
+			Type:               crdv1alpha1.PacketCaptureRunning,
+			Status:             metav1.ConditionStatus(v1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+		})
+	} else if status == packetCapturePhasePending || status == packetCapturePhaseUnknown {
+		conditions = append(conditions, crdv1alpha1.PacketCaptureCondition{
+			Type:               crdv1alpha1.PacketCapturePending,
+			Status:             metav1.ConditionStatus(v1.ConditionTrue),
+			LastTransitionTime: metav1.Now(),
+		})
+	} else {
+		return nil
+	}
+
+	toUpdate, getErr := c.packetCaptureLister.Get(name)
+	if getErr != nil {
+		return getErr
+	}
+	if retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		newCond := mergeConditions(toUpdate.Status.Conditions, conditions)
+		if conditionSliceEqualsIgnoreLastTransitionTime(toUpdate.Status.Conditions, newCond) {
+			return nil
+		}
+		toUpdate.Status.Conditions = newCond
+		klog.V(2).InfoS("Updating PacketCapture", "name", name, "status", toUpdate.Status)
+		_, updateErr := c.crdClient.CrdV1alpha1().PacketCaptures().UpdateStatus(context.TODO(), toUpdate, metav1.UpdateOptions{})
+		if updateErr != nil && apierrors.IsConflict(updateErr) {
+			var getErr error
+			if toUpdate, getErr = c.crdClient.CrdV1alpha1().PacketCaptures().Get(context.TODO(), name, metav1.GetOptions{}); getErr != nil {
+				return getErr
+			}
+		}
+		return updateErr
+	}); retryErr != nil {
+		return retryErr
+	}
+
+	return nil
+}
+
+func (c *Controller) updatePacketCaptureStatus(name string, num int32, path string, err error) error {
```

Comment:
Ideally we could have a single function to upload the status to the desired value, which would take all the state information required as a parameter